src/hotspot/share/classfile/lambdaFormInvokers.cpp

```diff
@@ -0,0 +1,156 @@
+  Klass*  cds_klass = SystemDictionary::resolve_or_null(cds_name, THREAD);
+  guarantee(cds_klass != NULL, "jdk/internal/misc/CDS must exist!");
+
+  int len = _lambdaform_lines->length();
+  objArrayHandle list_lines = oopFactory::new_objArray_handle(SystemDictionary::String_klass(), len, CHECK);
+  for (int i = 0; i < len; i++) {
+    char* record = _lambdaform_lines->at(i);
+    record += strlen(lambda_form_invoker_tag()) + 1; // skip the @lambda_form_invoker prefix
+    Handle h_line = java_lang_String::create_from_str(record, CHECK);
+    list_lines->obj_at_put(i, h_line());
+  }
+
+  //
+  // Object[] CDS.generateLambdaFormHolderClasses(String[] lines)
+  // the returned Object[] layout:
+  //   name, byte[], name, byte[] ....
+  Symbol* method = vmSymbols::generateLambdaFormHolderClasses();
+  Symbol* signrs = vmSymbols::generateLambdaFormHolderClasses_signature();
+
+  JavaValue result(T_OBJECT);
+  JavaCalls::call_static(&result, cds_klass, method, signrs, list_lines, THREAD);
+  objArrayHandle h_array(THREAD, (objArrayOop)result.get_jobject());
+  if (!HAS_PENDING_EXCEPTION) {
+    if (h_array() == NULL) {
+      log_info(cds)("Failed call to %s.%s", cds_name->as_C_string(), method->as_C_string());
+      return;
+    }
+  } else {
+    log_info(cds)("Exception happened: %s", PENDING_EXCEPTION->klass()->name()->as_C_string());
+    CLEAR_PENDING_EXCEPTION;
+    return;
+  }
+
+  int sz = h_array->length();
+  assert(sz % 2 == 0 && sz >= 2, "Must be even size of length");
+  for (int i = 0; i < sz; i+= 2) {
+    Handle h_name(THREAD, h_array->obj_at(i));
+    typeArrayHandle h_bytes(THREAD, (typeArrayOop)h_array->obj_at(i+1));
+    assert(h_name != NULL, "Class name is NULL");
+    assert(h_bytes != NULL, "Class bytes is NULL");
+
+    char *class_name = java_lang_String::as_utf8_string(h_name());
+    int len = h_bytes->length();
+    // make a copy of class bytes so GC will not affect us.
+    char *buf = resource_allocate_bytes(THREAD, len);
+    memcpy(buf, (char*)h_bytes->byte_at_addr(0), len);
+    ClassFileStream st((u1*)buf, len, NULL, ClassFileStream::verify);
+
+    reload_class(class_name, st, THREAD);
+    // free buf
+    resource_free_bytes(buf, len);
+
+    if (HAS_PENDING_EXCEPTION) {
+      log_info(cds)("Exception happened: %s", PENDING_EXCEPTION->klass()->name()->as_C_string());
+      log_info(cds)("Could not create InstanceKlass for class %s", class_name);
+      CLEAR_PENDING_EXCEPTION;
+      return;
+    }
+  }
+}
+
+// class_handle - the class name, bytes_handle - the class bytes
+void LambdaFormInvokers::reload_class(char* name, ClassFileStream& st, TRAPS) {
+  Symbol* class_name = SymbolTable::new_symbol((const char*)name);
+  // the class must exist
+  Klass* klass = SystemDictionary::resolve_or_null(class_name, THREAD);
+  assert(klass->is_instance_klass(), "Should be");
+  if (klass == NULL) {
+    log_info(cds)("Class %s not present, skip", name);
+    return;
+  }
```

Comment:
`assert(klass->is_instance_klass(), "Should be");` should be after the NULL check of `klass`